File: javascript/76_-_Minimum_Window_Substring.js
/**
 * 76. Minimum Window Substring - Hard
 * Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that
 * every character in t (including duplicates) is included in the window. If there is no such substring, return
 * the empty string "".
 * The testcases will be generated such that the answer is unique.
 * A substring is a contiguous sequence of characters within the string.
 */

/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    const freq = new Map();
    let ans = '';
    
    for (const c of t) {
        if (freq.has(c)) {
            freq.set(c, freq.get(c) + 1);
        } else {
            freq.set(c, 1);
        }
    }
    
    let begin = 0;
    let end = 0;
    let length = s.length + 1;
    let counter = freq.size;
    
    while (end < s.length) {
        const endChar = s.charAt(end);
        
        if (freq.has(endChar)) {
            freq.set(endChar, freq.get(endChar) - 1);
            
            if (freq.get(endChar) === 0) {
                --counter;
            }
        }
        
        ++end;
        
        while (counter === 0) {
            if (end - begin < length) {
                length = end - begin;
                ans = s.substring(begin, end);
            }
            
            const beginChar = s.charAt(begin);
            
            if (freq.has(beginChar)) {
                freq.set(beginChar, freq.get(beginChar) + 1);
                
                if (freq.get(beginChar) > 0) {
                    ++counter;
                }
            }
            
            ++begin;
        }
    }
    
    return ans;
};

File: csharp/76_-_Minimum_Window_Substring.cs
/**
 * 76. Minimum Window Substring - Hard
 * Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that
 * every character in t (including duplicates) is included in the window. If there is no such substring, return
 * the empty string "".
 * The testcases will be generated such that the answer is unique.
 * A substring is a contiguous sequence of characters within the string.
 */

public class Solution {
    public string MinWindow(string s, string t) {
        Dictionary<char,int> freq = new Dictionary<char,int>();
        
        foreach (char c in t) {
            if (freq.ContainsKey(c)) {
                freq[c] += 1;
            } else {
                freq[c] = 1;
            }
        }
        
        int begin = 0;
        int end = 0;
        int counter = freq.Count;
        int length = int.MaxValue;
        
        string ans = "";
        
        while (end < s.Length) {
            char endChar = s[end];
            
            if (freq.ContainsKey(endChar)) {
                freq[endChar] -= 1;
                
                if (freq[endChar] == 0) {
                    --counter;
                }
            }
            
            ++end;
            
            while (counter == 0) {
                if (end - begin < length) {
                    length = end - begin;
                    ans = s.Substring(begin, end - begin);
                }
                
                char startChar = s[begin];
                
                if (freq.ContainsKey(startChar)) {
                    freq[startChar] += 1;
                    
                    if (freq[startChar] > 0) {
                        ++counter;
                    }
                }
                
                ++begin;
            }
        }
        
        return ans;
    }
}